mobile_verifier/src/speedtests_average.rs

```diff
@@ -0,0 +1,478 @@
+use crate::speedtests::{self, Speedtest};
+use chrono::{DateTime, Duration, Utc};
+use file_store::{file_sink::FileSinkClient, traits::TimestampEncode};
+use helium_crypto::PublicKeyBinary;
+use helium_proto::services::poc_mobile::{self as proto, SpeedtestAvgValidity};
+use rust_decimal::Decimal;
+use rust_decimal_macros::dec;
+use std::collections::HashMap;
+
+pub const SPEEDTEST_LAPSE: i64 = 48;
+const MIN_DOWNLOAD: u64 = mbps(30);
+const MIN_UPLOAD: u64 = mbps(2);
+const MAX_LATENCY: u32 = 100;
+pub const MIN_REQUIRED_SAMPLES: usize = 2;
+
+pub type EpochAverages = HashMap<PublicKeyBinary, SpeedtestAverage>;
+
+#[derive(Debug, Clone)]
+pub struct SpeedtestAverage {
+    pub pubkey: PublicKeyBinary,
+    pub window_size: usize,
+    pub upload_speed_avg_bps: u64,
+    pub download_speed_avg_bps: u64,
+    pub latency_avg_ms: u32,
+    pub validity: SpeedtestAvgValidity,
+    pub reward_multiplier: Decimal,
+}
+
+impl<'a, I: ?Sized> From<&'a I> for SpeedtestAverage
+where
+    &'a I: IntoIterator<Item = &'a Speedtest>,
+{
+    fn from(iter: &'a I) -> Self {
+        let mut id = vec![]; // eww!
+        let mut window_size = 0;
+        let mut sum_upload = 0;
+        let mut sum_download = 0;
+        let mut sum_latency = 0;
+
+        for Speedtest { report, .. } in
+            speedtests_without_lapsed(iter.into_iter(), Duration::hours(SPEEDTEST_LAPSE))
+        {
+            id = report.pubkey.as_ref().to_vec(); // eww!
```

Comment:
Instead of just overwriting the pubkey/id each iteration can you set it once based on the first element of the loop or else pass it in as an additional argument and then filter in the iterator loop based on each speedtest having the same pubkey?

Reply:
here's what i was thinking:
```
impl<'a, I: ?Sized> From<&'a I> for SpeedtestAverage
where
    &'a I: IntoIterator<Item = &'a Speedtest>,
{
    fn from(iter: &'a I) -> Self {
        let mut window_size = 0;
        let mut sum_upload = 0;
        let mut sum_download = 0;
        let mut sum_latency = 0;

        let mut speedtest_iter = speedtests_without_lapsed(iter.into_iter(), Duration::hours(SPEEDTEST_LAPSE)).peekable();
        let id = speedtest_iter
            .peek()
            .map(|Speedtest { report, .. }| report.pubkey.as_ref().to_vec()).unwrap_or_default();
        for Speedtest { report, .. } in speedtest_iter 
        {
            if report.pubkey.as_ref().to_vec() == id {
                sum_upload += report.upload_speed;
                sum_download += report.download_speed;
                sum_latency += report.latency;
                window_size += 1;
            }
        }

        if window_size > 0 {
            let upload_speed_avg_bps = sum_upload / window_size as u64;
            let download_speed_avg_bps = sum_download / window_size as u64;
            let latency_avg_ms = sum_latency / window_size as u32;
            let validity = validity(
                window_size as usize,
                upload_speed_avg_bps,
                download_speed_avg_bps,
                latency_avg_ms,
            );
            let tier = SpeedtestTier::new(
                window_size as usize,
                upload_speed_avg_bps,
                download_speed_avg_bps,
                latency_avg_ms,
            );
            let reward_multiplier = tier.into_multiplier();
            SpeedtestAverage {
                pubkey: id.into(),
                window_size: window_size as usize,
                upload_speed_avg_bps,
                download_speed_avg_bps,
                latency_avg_ms,
                validity,
                reward_multiplier,
            }
        } else {
            SpeedtestAverage {
                pubkey: id.into(),
                window_size: 0,
                upload_speed_avg_bps: sum_upload,
                download_speed_avg_bps: sum_download,
                latency_avg_ms: sum_latency,
                validity: proto::SpeedtestAvgValidity::TooFewSamples,
                reward_multiplier: Decimal::ZERO,
            }
        }
    }
}
```
logically this works and it compiles. i'm not a fan of the `unwrap_or_default` since that _theoretically_ could result in aggregating the speedtest average under a `SpeedtestAverage` with an empty Vec<u8> as the `Pubkey` but since that seems really unlikely that we'd have a valid `CellSpeedtest` with an empty pubkey than you should be safe to just replace this with an `unwrap()`.

i like the `peekable` implementation on the iterator to grab the first element of the list without consuming it and then checking subsequent members of the list to ensure the pubkeys match before including them in the average, but that's probably unnecessary and if overwriting the id on each iteration of the for loop over the collection of CellSpeedtests isn't an expensive operation this is probably unnecessary so i'll leave it to you to take or ignore